torch/distributed/device_mesh.py

```diff
@@ -89,7 +89,11 @@ def create_child_mesh(
                     res_sub_mesh = sub_mesh
 
             res_sub_mesh._dim_group_infos = [device_mesh._dim_group_infos[mesh_dim]]  # type: ignore[possibly-undefined]
+            # Update the hash of the submesh to take the parent mesh into consideration
+            # so the same mesh with or without the parent mesh is considered different.
+            res_sub_mesh._hash = hash((res_sub_mesh._hash, device_mesh._hash))
```

Comment:
If you update the submesh hash to include the parent mesh hash. You would need to change __eq__ to match.

A overall restriction we have to ensure here is that hash and eq must match. So if the mesh have a parent mesh, we would have to compare the parent mesh in `__eq__`